### PR TITLE
Merge dev into main

### PR DESCRIPTION
**Pull request by ttqureshi**
### To get the weather extremes of a year: `python3 main.py ./ -e 2012`:
**Output:**
<==== Report for the year 2012: ====>
Highest: 32C on June 21
Lowest: -4C on January 10
Humidity: 100% on January 18

### To get the average stats of a month: `python3 main.py ./ -a 2009/4`
**Output:**
<==== Report for the month: April 2009 ====>
Highest Average: 17.2C
Lowest Average: 10.7C
Average Mean Humidity: 44.0%

### To get the daily temperature bar charts of a month: `python3 main.py ./ -c 2010/7`
**Output:**
<==== Temperature Bar Charts for July 2010 ====>
1 +++++++++++++++++++++++++++ 27C
1 +++++++++++++++++ 17C
2 ++++++++++++++++++++++++ 24C
2 ++++++++++++++++++ 18C
3 ++++++++++++++++++++++++++ 26C
3 ++++++++++++++++++ 18C
4 ++++++++++++++++++++++++ 24C
4 +++++++++++++++ 15C
5 +++++++++++++++++++++++++ 25C
5 ++++++++++++++++ 16C
6 +++++++++++++++++++++++ 23C
6 ++++++++++++++++ 16C
7 +++++++++++++++++++++++ 23C
7 ++++++++++++++++ 16C
...

### To get highest and lowest temperature bar chart of a day in one line: `python3 main.py ./ -c 2010/7 --inline`
*Output:*
<==== Temperature Bar Charts for July 2010 ====>
1 +++++++++++++++++ +++++++++++++++++++++++++++ 17C - 27C
2 ++++++++++++++++++ ++++++++++++++++++++++++ 18C - 24C
3 ++++++++++++++++++ ++++++++++++++++++++++++++ 18C - 26C
4 +++++++++++++++ ++++++++++++++++++++++++ 15C - 24C
5 ++++++++++++++++ +++++++++++++++++++++++++ 16C - 25C
6 ++++++++++++++++ +++++++++++++++++++++++ 16C - 23C
7 ++++++++++++++++ +++++++++++++++++++++++ 16C - 23C
8 ++++++++++++++++++ +++++++++++++++++++++++++++ 18C - 27C
9 +++++++++++++++++++ ++++++++++++++++++++++++++ 19C - 26C
10 ++++++++++++++++++++ +++++++++++++++++++++++++ 20C - 25C
...

### Multiple reports at a time: `python3 main.py ./ -c 2010/7 -a 2009/4 -e 2012 --inline` 
*Output:*
<==== Report for the year 2012: ====>
Highest: 32C on June 21
Lowest: -4C on January 10
Humidity: 100% on January 18

<==== Report for the month: April 2009 ====>
Highest Average: 17.2C
Lowest Average: 10.7C
Average Mean Humidity: 44.0%

<==== Temperature Bar Charts for July 2010 ====>
1 +++++++++++++++++ +++++++++++++++++++++++++++ 17C - 27C
2 ++++++++++++++++++ ++++++++++++++++++++++++ 18C - 24C
3 ++++++++++++++++++ ++++++++++++++++++++++++++ 18C - 26C
4 +++++++++++++++ ++++++++++++++++++++++++ 15C - 24C
5 ++++++++++++++++ +++++++++++++++++++++++++ 16C - 25C
6 ++++++++++++++++ +++++++++++++++++++++++ 16C - 23C
7 ++++++++++++++++ +++++++++++++++++++++++ 16C - 23C
8 ++++++++++++++++++ +++++++++++++++++++++++++++ 18C - 27C
9 +++++++++++++++++++ ++++++++++++++++++++++++++ 19C - 26C
10 ++++++++++++++++++++ +++++++++++++++++++++++++ 20C - 25C
...

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
-weatherfiles.zip
-Weatherman.pdf
\ No newline at end of file
+Weatherman.pdf
+rough/
+weatherfiles/
+main.py
```
Comment (irtazaakram) [inline .gitignore:4]: let's add a blank line at the end of file.

File: weatherman.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
+import argparse
```
Comment (irtazaakram) [inline weatherman.py:1]: Let's refactor this code to make use of classes for calculations and reporting.

**Review by irtazaakram — COMMENTED** [review]
Lets avoid pushing `.DS_Store` & `weatherfiles.zip` to repo

**Comment by ttqureshi** [comment]
> Implemented parser to parse files and save weather readings

**Comment by ttqureshi** [comment]
> Lets avoid pushing `.DS_Store` & `weatherfiles.zip` to repo

okay

File: weatherman.py (hunk 2 above)
Comment (ttqureshi) [reply weatherman.py:1]: alright

File: code_files/file_parser.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                    reading = WeatherReading(
+                        date=row[list(row.keys())[0]],
+                        max_temp=int(row['Max TemperatureC']),
+                        min_temp=int(row['Min TemperatureC']),
+                        max_humidity=int(row['Max Humidity']),
+                        mean_humidity=int(row[' Mean Humidity'])
+                    )
+                    self.weather_readings.append(reading)
+                except ValueError:
+                    continue
```
Comment (irtazaakram) [inline code_files/file_parser.py:46]: Let's move these calculations to `WeatherReading.validate_reading` and avoid using try except through out whole of this task.

File: code_files/main.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import argparse
```
Comment (irtazaakram) [inline code_files/main.py:1]: Let's move this main file out of this folder (code_files) .

Comment (irtazaakram) [inline code_files/main.py:1]: `main.py` --> `weatherman.py `

File: code_files/main.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+from file_parser import Parser
```
Comment (irtazaakram) [inline code_files/main.py:2]: `Parser` --> `WeatherParser`

File: code_files/weather_reporter.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        h_temp, l_temp, h_humidity = Computations.compute_extreme_stats(readings)
+        date_htemp = h_temp.date.split("-")
+        date_ltemp = l_temp.date.split("-")
+        date_hhumid = h_humidity.date.split("-")
+
+        print(f"\n<==== Report for the year {date_htemp[0]}: ====>")
+        print(f"Highest: {h_temp.max_temp}C on {MONTHS[int(date_htemp[1])]} {date_htemp[2]}")
+        print(f"Lowest: {l_temp.min_temp}C on {MONTHS[int(date_ltemp[1])]} {date_ltemp[2]}")
+        print(f"Humidity: {h_humidity.max_humidity}% on {MONTHS[int(date_hhumid[1])]} {date_hhumid[2]}")
+        print("-------------------------------------")
```
Comment (irtazaakram) [inline code_files/weather_reporter.py:32]: Let's only use `WeatherReporter` for printing only. All calculations should be moved to calculations class.

File: code_files/weather_reporter.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        print(f"Lowest Average: {avg_min_temp:.1f}C")
+        print(f"Average Mean Humidity: {avg_mean_humidity:.1f}%")
+        print("-------------------------------------")
+
+    def generate_report_barchart(year, month, readings, isInline):
+        """Generates bar chart of daily highest and lowest temperatures for a given month of the year"""
+
+        RED = "\033[91m"
+        BLUE = "\033[94m"
+        RESET = "\033[0m"
```
Comment (irtazaakram) [inline code_files/weather_reporter.py:50]: Let's make a separate class for these constants.

File: code_files/computations.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
+class Computations:
```
Comment (irtazaakram) [inline code_files/computations.py:1]: `Computations` --> `ReportCalculator`

**Review by irtazaakram — COMMENTED** [review]
Let's go through pep8 one more time and try to follow it throughout the codebase.

**Comment by ttqureshi** [comment]
> Let's go through pep8 one more time and try to follow it throughout the codebase.

ok

File: code_files/file_parser.py (hunk 3 above)
Comment (ttqureshi) [reply code_files/file_parser.py:46]: sure

File: code_files/main.py (hunk 4 above)
Comment (ttqureshi) [reply code_files/main.py:1]: okay

**Comment by ttqureshi** [comment]
I have made the suggested changes. Please review them.

File: weatherman.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+from code_files.weather_parser import WeatherParser
+from code_files.weather_reporter import WeatherReporter
+from code_files.report_calculator import ReportCalculator
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("extract_to", type=str)
+    parser.add_argument("-e", "--extremes", type=int)
+    parser.add_argument("-a", "--averages", type=str)
+    parser.add_argument("-c", "--chart", type=str)
```
Comment (irtazaakram) [inline weatherman.py:11]: Can we use `datetime` type instead of `int` here

File: weatherman.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
+import argparse
+from code_files.weather_parser import WeatherParser
+from code_files.weather_reporter import WeatherReporter
+from code_files.report_calculator import ReportCalculator
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("extract_to", type=str)
```
Comment (irtazaakram) [inline weatherman.py:8]: Let's also add help messages for users.

File: weatherman.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    parser.add_argument("-e", "--extremes", type=int)
```
Comment (irtazaakram) [inline weatherman.py:9]: Let's also come up with more useful name args e.g `-e` & `--yearly_report` instead of `--extremes`

File: weatherman.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    parser.add_argument("-a", "--averages", type=str)
```
Comment (irtazaakram) [inline weatherman.py:10]: Let's make it more useful by allowing multiple dates to be passed.

File: weatherman.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    parser.extract_zip()
+    parser.parse_all_files()
+
+    if args.extremes:
+        yearly_readings = [
+            r for r in parser.weather_readings if r.date.startswith(str(args.extremes))
+        ]
+        if yearly_readings:
+            weather_extremes = ReportCalculator.compute_extreme_stats(yearly_readings)
+            WeatherReporter.generate_report_extremes(weather_extremes)
```
Comment (irtazaakram) [inline weatherman.py:26]: We are using `ReportCalculator` & `WeatherReporter` classes as a functions here. Let's make sure we are following oop approach.

File: weatherman.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+        if monthly_readings:
+            for reading in monthly_readings:
+                if reading.max_temp and reading.min_temp:
+                    high_bar = "+" * reading.max_temp
+                    low_bar = "+" * reading.min_temp
+                    date = reading.date.split("-")
+                    WeatherReporter.generate_report_barchart(
+                        date, high_bar, low_bar, args.inline
+                    )
```
Comment (irtazaakram) [inline weatherman.py:59]: Let's not do any formatting in main function.

File: code_files/main.py (hunk 4 above)
Comment (ttqureshi) [reply code_files/main.py]: [See this latest commit](https://github.com/ttqureshi/WeatherMan/commit/a9bfba76030779f62875a46887178ea2565ca521)

File: code_files/main.py (hunk 5 above)
Comment (ttqureshi) [reply code_files/main.py]: [Resolved in this commit](https://github.com/ttqureshi/WeatherMan/commit/a9bfba76030779f62875a46887178ea2565ca521)

File: code_files/weather_reporter.py (hunk 6 above)
Comment (ttqureshi) [reply code_files/weather_reporter.py]: [Latest commit](https://github.com/ttqureshi/WeatherMan/commit/a9bfba76030779f62875a46887178ea2565ca521)

File: weatherman.py (hunk 9 above)
Comment (ttqureshi) [reply weatherman.py]: [Using datetime](https://github.com/ttqureshi/WeatherMan/pull/1/commits/b1f61ff2496d9efa085df59157f39ad21df4b856)

File: weatherman.py (hunk 10 above)
Comment (ttqureshi) [reply weatherman.py]: [Added help messages](https://github.com/ttqureshi/WeatherMan/pull/1/commits/90150ed551723c3fabc39a9b62a6234e3180bc01)

File: weatherman.py (hunk 11 above)
Comment (ttqureshi) [reply weatherman.py]: [Meaning argument names](https://github.com/ttqureshi/WeatherMan/pull/1/commits/b6d935340b0b0850b476f6ec2ed11e67d57d5fe8)

File: weatherman.py (hunk 12 above)
Comment (ttqureshi) [reply weatherman.py]: [Allowing multiple dates to be passed](https://github.com/ttqureshi/WeatherMan/pull/1/commits/6f2c4b5a1bdecc27fab75a9c7b9eb714a6baad70)

File: weatherman.py (hunk 13 above)
Comment (ttqureshi) [reply weatherman.py]: [Refactoring code to make use of OOP for `ReportCalculator` & `WeatherReporter` classes](https://github.com/ttqureshi/WeatherMan/pull/1/commits/4b32012e0123a6b56cffad28fac085e86fb58a6e)

File: weatherman.py (hunk 14 above)
Comment (ttqureshi) [reply weatherman.py]: [Formatting](https://github.com/ttqureshi/WeatherMan/pull/1/commits/29f7b39b1326faaabd99d94138385d89417c20ab)

**Comment by ttqureshi** [comment]
I have made changes as per your suggestions. Kindly review them.

File: code_files/report_calculator.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+class ReportCalculator:
+    def __init__(self, readings):
+        self.readings = readings
+    
+    def compute_extreme_stats(self):
+        """Computes the extreme weather statistics for a particular year"""
+        
+        highest_temp_reading = max(
+            self.readings,
+            key=lambda r: r.max_temp if r.max_temp is not None else float("-inf"),
```
Comment (irtazaakram) [inline code_files/report_calculator.py:13]: Do we need this check now that we are not parsing empty rows?

Comment (ttqureshi) [reply code_files/report_calculator.py:13]: This check is required. There are certain rows in which some values are missing, and they are read as `None`. If this check is removed, we get the following error:
```
Traceback (most recent call last):
  File "/Users/tayyab.tahir/Documents/MTTQ/Training/Task1_WeatherMan/weatherman.py", line 27, in <module>
    weather_extremes = report_calculator.compute_extreme_stats()
  File "/Users/tayyab.tahir/Documents/MTTQ/Training/Task1_WeatherMan/code_files/report_calculator.py", line 11, in compute_extreme_stats
    highest_temp_reading = max(
TypeError: '>' not supported between instances of 'NoneType' and 'int'
tayyab.tahir@A006-01397 Task1_WeatherMan % python3
```

File: code_files/weather_parser.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
+import os
+import csv
+import zipfile
+from code_files.weather_readings import WeatherReading
```
[the remark this answers is not in the document]
Comment (ttqureshi) [reply code_files/weather_parser.py]: [pep8 import order](https://github.com/ttqureshi/WeatherMan/pull/1/commits/f57eab6f6a8ba30a7475b5d0b20393fe77ce22a2)

File: weatherman.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
+import argparse
+from datetime import datetime
+from code_files.weather_parser import WeatherParser
+from code_files.weather_reporter import WeatherReporter
+from code_files.report_calculator import ReportCalculator
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("extract_to", type=str, help='Directory to extract the results to.')
+    parser.add_argument("-e", "--yearly_report", type=lambda e: datetime.strptime(e, '%Y'), nargs='+', help='Extreme weather stats for a year (format: YYYY).')
```
[the remark this answers is not in the document]
Comment (ttqureshi) [reply weatherman.py]: [Fix argparse error](https://github.com/ttqureshi/WeatherMan/pull/1/commits/bf1e860d1d70e5d2024b30b6b83a31596f30d83e)

File: weatherman.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    if args.yearly_report:
+        for yearly_report in args.yearly_report:
+            yearly_readings = [r for r in parser.weather_readings if r.date.startswith(yearly_report.strftime('%Y'))]
+            
+            if yearly_readings:
+                report_calculator = ReportCalculator(yearly_readings)
+                weather_extremes = report_calculator.compute_extreme_stats()
+
+                weather_reporter = WeatherReporter(weather_extremes)
+                weather_reporter.generate_report_extremes()
```
[the remark this answers is not in the document]
Comment (ttqureshi) [reply weatherman.py]: [One time object instantiation](https://github.com/ttqureshi/WeatherMan/pull/1/commits/f6bfc70bc9fd2f971492129b2c4575b8f0862a71)

File: weatherman.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
+import argparse
+from datetime import datetime
+from code_files.weather_parser import WeatherParser
+from code_files.weather_reporter import WeatherReporter
+from code_files.report_calculator import ReportCalculator
```
[the remark this answers is not in the document]
Comment (ttqureshi) [reply weatherman.py]: [Organizing Imports](https://github.com/ttqureshi/WeatherMan/pull/1/commits/5e94679843292972eafbee834c3ecd05f888727d)

**Comment by ttqureshi** [comment]
Suggested Changes have been made. Please review them

**Pull request closed without merging**